src/chess/bitboard.h

```diff
@@ -190,17 +190,29 @@ class Move {
  public:
   enum class Promotion : std::uint8_t { None, Queen, Rook, Bishop, Knight };
   Move() = default;
-  Move(BoardSquare from, BoardSquare to) : from_(from), to_(to) {}
-  Move(BoardSquare from, BoardSquare to, Promotion promotion)
-      : from_(from), to_(to), promotion_(promotion) {}
+  Move(BoardSquare from, BoardSquare to) {
+    data_ = 0;
+    data_ |= to.as_int();
+    data_ |= from.as_int() << 6;
+  }
+  Move(BoardSquare from, BoardSquare to, Promotion promotion){
+    data_ = 0;
+    data_ |= to.as_int();
+    data_ |= from.as_int() << 6;
+    data_ |= (uint8_t)promotion << 12;
+  }
   Move(const std::string& str, bool black = false);
   Move(const char* str, bool black = false) : Move(std::string(str), black) {}
 
-  BoardSquare from() const { return from_; }
-  BoardSquare to() const { return to_; }
-  Promotion promotion() const { return promotion_; }
-  bool IsCastling() const { return castling_; }
-  void SetCastling() { castling_ = true; }
+  BoardSquare to() const      { return (BoardSquare)((data_ & 0b0000000000111111)); }
+  BoardSquare from() const    { return (BoardSquare)((data_ & 0b0000111111000000) >> 6); }
+  Promotion promotion() const { return (Promotion)  ((data_ & 0b0111000000000000) >> 12); }
+  bool IsCastling() const     { return (bool)       ((data_ & 0b1000000000000000) >> 15); }
+  void SetCastling() { data_ |= 0b1000000000000000; }
+
+  void SetTo(BoardSquare to){ data_ =             (data_ & 0b1111111111000000) | to.as_int(); }
+  void SetFrom(BoardSquare from){ data_ =         (data_ & 0b11110000001111111) | (from.as_int() << 6); }
```

Comment:
Maybe we should add a comment to the BoardSquare class that its value is guaranteed to have only the lowest 6 bits set?

Reply:
Done